Translate from regex to recognizer combinators

raphlinus opened the pull request:
Gets rid of the dependency on regex and uses recognizer combinators.
The recognizers are just included, rather than having a dependency on
a separate crate, as they're small and simple enough.

Also adds some tests to make sure numeric overflow is handled without
panic (the previous version would not pass these).

raphlinus committed:
Translate from regex to recognizer combinators

Gets rid of the dependency on regex and uses recognizer combinators.
The recognizers are just included, rather than having a dependency on
a separate crate, as they're small and simple enough.

Also adds some tests to make sure numeric overflow is handled without
panic (the previous version would not pass these).

steveklabnik commented on src/recognize.rs:
```diff
@@ -0,0 +1,135 @@
+// Copyright 2017 Google Inc. All rights reserved.
```
how does "all rights reserved" work with the MIT/Apache license here?

raphlinus replied:
The intent is for it to be covered by the same licenses as the repo. I think the best way to do this (while preserving the Google copyright for the new file) is to just add boilerplate so there's no confusion.

steveklabnik replied:
Great! I just wanted to make sure.

steveklabnik reviewed (commented):
This is amazing raph, thank you so much! I can't say enough good things. I have one question about the license here, but other than that, let's :shipit:

raphlinus committed:
Add license boilerplate for new file

License boilerplate consists of standard MIT and Apache boilerplate,
preceded by a simple statement saying it's either at your choice.

raphlinus committed:
Merge branch 'master' of https://github.com/steveklabnik/semver-parser

steveklabnik commented:
🎊

kamalmarhubi commented:
@raphlinus thanks so much for this! It's been bugging me for almost a year since I opened a couple of PRs on rustc_version!
